src/datatype.c

```diff
@@ -147,12 +196,15 @@ static jl_datatype_layout_t *jl_get_layout(uint32_t nfields,
         }
     }
 
-    // allocate a new descriptor
-    // TODO: lots of these are the same--take advantage of the fact these are immutable to combine them
-    uint32_t fielddesc_size = jl_fielddesc_size(fielddesc_type);
-    jl_datatype_layout_t *flddesc = (jl_datatype_layout_t*)jl_gc_perm_alloc(
-                sizeof(jl_datatype_layout_t) + nfields * fielddesc_size + (npointers << fielddesc_type),
-                0, 4, 0);
+    // allocate a new descriptor, on the stack if possible.
+    size_t fields_size = nfields * jl_fielddesc_size(fielddesc_type);
+    size_t pointers_size = (npointers << fielddesc_type);
+    size_t flddesc_sz = sizeof(jl_datatype_layout_t) + fields_size + pointers_size;
+    int should_malloc = flddesc_sz >= jl_page_size;
+    jl_datatype_layout_t* mallocmem = (jl_datatype_layout_t*)(should_malloc ? malloc(flddesc_sz) : NULL);
+    jl_datatype_layout_t* allocamem = (jl_datatype_layout_t*)(should_malloc ? NULL : alloca(flddesc_sz));
+    jl_datatype_layout_t *flddesc = should_malloc ? mallocmem : allocamem;
+    assert(flddesc);
```

Comment:
style nit: seems a little unnecessarily verbose, although we don't pay by the line
```suggestion
    jl_datatype_layout_t *flddesc = (jl_datatype_layout_t*)(should_malloc ? malloc(flddesc_sz) : alloca(flddesc_sz));
```